data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/HashRing.java

```diff
@@ -0,0 +1,64 @@
+package com.amazon.dataprepper.plugins.processor.peerforwarder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.zip.CRC32;
+import java.util.zip.Checksum;
+
+@NotThreadSafe
+public class HashRing {
+    private final List<String> serverIps = new ArrayList<>();
+    private final int numVirtualNodes;
+    private final TreeMap<Long, String> virtualNodes = new TreeMap<>();
+
+    public HashRing(@Nonnull final List<String> serverIps, final int numVirtualNodes) {
+        this.numVirtualNodes = numVirtualNodes;
+        final List<String> sortedServerIps = new ArrayList<>(serverIps);
+        Collections.sort(sortedServerIps);
+        for (final String serverIp: sortedServerIps) {
+            addServerIp(serverIp);
+        }
+    }
+
+    public List<String> getServerIps() {
+        return serverIps;
+    }
+
+    private void addServerIp(@Nonnull final String serverIp) {
+        serverIps.add(serverIp);
+        final byte[] serverIpInBytes = serverIp.getBytes();
+        final Checksum crc32 = new CRC32();
+        final ByteBuffer intBuffer = ByteBuffer.allocate(4);
+        for (int i = 0; i < numVirtualNodes; i++) {
+            crc32.update(serverIpInBytes, 0, serverIpInBytes.length);
+            intBuffer.putInt(i);
+            crc32.update(intBuffer.array(), 0, intBuffer.array().length);
+            virtualNodes.putIfAbsent(crc32.getValue(), serverIp);
```

Comment:
Actually the token space size is [0, 2**32] for CRC32 (notice that it returns long type). I expect rare collision in that range. Otherwise, we might replace it with some 64 or 128 bit long hashing algorithm

>  Are we missing out by not having a bounded token space size?

Not sure what "missing out" means here. In theory there is no need to manually put a bound b/c in the mental model it is a loop so the next element to maximum is the minimum.

Reply:
I'm not sure what we could be missing out on either ;) 

I've seen implementations where the token space size is bounded w/ a mod function, was wondering if having a "sparsely" populated ring has any downsides that I'm not aware of. Maybe not.

Reply:
I think sparsity might not be the issue. Instead, it is more about the evenness in the distribution of hashing results. We can try more randomized value to be hashed by MD5 if it turns out to be too biased.

Reply:
Sure, we can revisit later, no need to optimize now.

Related, the 2007 DDB paper discusses partitioning strategies: https://www.allthingsdistributed.com/files/amazon-dynamo-sosp2007.pdf